### PR TITLE
[feature] HTTP Proxy

### PR DESCRIPTION
**Pull request by EdGeraghty**
### Types
- [ ] Fix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Refactor (non-breaking change which improves code quality - QA thoroughly )
- [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)

#### 🔮 Features
Allow the specifying of an HTTP proxy
### QA

- [ ] Signup
- [x] Login
- [ ] Logout
- [ ] Unencrypted Chat 
- [ ] Encrypted Chat 
- [ ] Group Chats
- [ ] Profile Changes
- [ ] Theming Changes
- [ ] Force Kill and Restart

### Final Checklist
 
- [x] All associated issues have been linked to PR
- [ ] All necessary changes made to the documentation
- [x] Parties interested in these changes have been notified

### CODE REVIEW

**Comment by EdGeraghty** [comment]
Starting work on #415

**Comment by EdGeraghty** [comment]
This code needs ` TextEditingController ` to be patched, but I believe otherwise works

File: lib/global/libs/matrix/media.dart [hunk 1]
```diff
@@ -18,8 +18,6 @@ class MatrixMedia {
     final String? serverName = params['serverName'];
     final String mediaUri = params['mediaUri'];
 
-    httpClient = createClient();
```
Comment (ereio) [inline lib/global/libs/matrix/media.dart:21]: These are necessary to run these functions in a background thread. Each isolate in Dart has no context or shared RAM from another, similar to NodeJS. They're not "threads" proper.

File: lib/global/libs/matrix/media.dart [hunk 2]
```diff
@@ -75,8 +73,6 @@ class MatrixMedia {
...
```
Comment (ereio) [inline lib/global/libs/matrix/media.dart:78]: note previous comment

File: lib/global/libs/matrix/rooms.dart [hunk 3]
```diff
@@ -67,8 +67,6 @@ abstract class Rooms {
     final int? timeout = params['timeout'];
     final String? filter = params['filter'];
 
-    httpClient = createClient();
```
Comment (ereio) [inline lib/global/libs/matrix/rooms.dart:70]: note previous comment

File: lib/global/libs/matrix/events.dart [hunk 4]
```diff
@@ -71,8 +71,6 @@ abstract class Events {
   ///
   /// https://matrix.org/docs/spec/client_server/latest#id251
   static Future<dynamic> fetchMessageEventsThreaded(Map params) async {
-    httpClient = createClient();
```
Comment (ereio) [inline lib/global/libs/matrix/events.dart:74]: note previous comment

**Review by ereio — CHANGES_REQUESTED** [review]

File: lib/views/intro/login/login-screen.dart [hunk 5]
```diff
@@ -281,6 +282,15 @@ class LoginScreenState extends State<LoginScreen> with Lifecycle<LoginScreen> {
                 },
               ),
             ),
+            IconButton(
+              icon: Icon(Icons.vpn_lock),
```
Comment (ereio) [inline lib/views/intro/login/login-screen.dart:286]: Change to the overflow (...) icon instead, as we'll likely be placing out-of-band SMS here as well

File: lib/store/settings/state.dart [hunk 6]
```diff
@@ -7,6 +7,7 @@ import 'package:syphon/store/settings/notification-settings/model.dart';
 import 'package:syphon/store/settings/theme-settings/model.dart';
 
 import './chat-settings/model.dart';
+import 'http-proxy-settings/model.dart';
```
Comment (ereio) [inline lib/store/settings/state.dart:10]: Rename `HttpProxySettings` to just `ProxySettings` and the directory structure equivalent in kebab case `proxy-settings`

**Review by ereio — CHANGES_REQUESTED** [review]

File: lib/global/libs/matrix/media.dart (hunk 1 above)
Comment (EdGeraghty) [reply lib/global/libs/matrix/media.dart]: Reverted the two referred commits

File: lib/global/libs/matrix/media.dart (hunk 2 above)
Comment (EdGeraghty) [reply lib/global/libs/matrix/media.dart]: reverted

File: lib/global/libs/matrix/rooms.dart (hunk 3 above)
Comment (EdGeraghty) [reply lib/global/libs/matrix/rooms.dart]: reverted

File: lib/global/libs/matrix/events.dart (hunk 4 above)
Comment (EdGeraghty) [reply lib/global/libs/matrix/events.dart]: reverted

File: lib/views/intro/login/login-screen.dart (hunk 5 above)
Comment (EdGeraghty) [reply lib/views/intro/login/login-screen.dart:286]: I went for the vertical one , as I think it looks better. Feel free to replace with horizontal

**Review by ereio — APPROVED** [review]
Everything looks good here code wise. Going to QA and fix the dialog issue, and if nothing else shows up, it's good to go.

File: lib/store/settings/proxy-settings/model.dart [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    host,
+    port,
+  }) =>
+      ProxySettings(
+        enabled: enabled ?? this.enabled,
+        host: host ?? this.host,
+        port: port ?? this.port,
+      );
+
+  Map<String, dynamic> toJson() => _$HttpProxySettingsToJson(this);
```
Comment (ereio) [inline lib/store/settings/proxy-settings/model.dart:36]: Make sure to rename both the `*Json()` functions from `HttpProxySettings*` to `ProxySettings*`

**Review by ereio — CHANGES_REQUESTED** [review]

File: lib/store/settings/proxy-settings/model.dart (hunk 7 above)
Comment (ereio) [reply lib/store/settings/proxy-settings/model.dart:36]: Sorry, testing this PR now and caught this after the fact

Comment (EdGeraghty) [reply lib/store/settings/proxy-settings/model.dart]: Done. I missed them too.

**Comment by ereio** [comment]
added a small PR to your fork that fixes the dialog inputs. Feel free to merge or re-implement the changes. I'll merge once  those are in. looks good!

**Comment by EdGeraghty** [comment]
> added a small PR to your fork that fixes the dialog inputs. Feel free to merge or re-implement the changes. I'll merge once those are in. looks good!

Done!

Thanks!

**Review by ereio — APPROVED** [review]

**Pull request merged**